rat/ops.py

```diff
@@ -24,77 +47,196 @@ def __str__(self):
 
 @dataclass
 class RealConstant(Expr):
+    """
+    Elementary value, cannot be folded anymore
+    """
+
     value: float
 
-    def code(self):
+    def code(self, scalar = False):
         if self.value == float("inf"):
             return "float('inf')"
         elif self.value == float("-inf"):
             return "float('-inf')"
         return f"{self.value}"
 
+    def __post_init__(self):
+        self.out_type = types.RealType
+
+    def fold(self):
+        return self
+
     def __str__(self):
         return f"RealConstant({self.value})"
 
 
 @dataclass
 class IntegerConstant(Expr):
+    """
+    Elementary value, cannot be folded anymore
+    """
+
     value: int
 
-    def code(self):
+    def code(self, scalar = False):
         return f"{self.value}"
 
+    def __post_init__(self):
+        self.out_type = types.IntegerType
+
+    def fold(self):
+        return self
+
     def __str__(self):
         return f"IntegerConstant({self.value})"
 
 
 @dataclass
 class Subscript(Expr):
+    """
+    Elementary value, cannot be folded anymore
+    """
+
     names: Tuple[str]
-    shifts: Tuple[Union[str, None]] = (None,)
+    shifts: Tuple[Union[int, None]] = (None,)
     variable: variables.SubscriptUse = None
 
     def get_key(self):
         return self.names
 
-    def code(self):
-        return f"subscripts['{self.variable.code()}']"
+    def code(self, scalar = False):
+        if scalar:
+            return f"subscripts['{self.variable.code()}'][index]"
```

Comment:
All the ops now take an extra argument `scalar` which indicates if they should do code generation assuming they're being evaluated like scalars or not.

If they're scalars, then assume the variable `index` is the scalar index.